hw/ip/otbn/data/bignum-insns.yml

```
@@ -393,7 +393,7 @@
   decode: *bn-sub-decode
   operation: |
     b_shifted = ShiftReg(b, st, sb)
-    (result, flags_out) = AddWithCarry(a, -b_shifted, ~FLAGS[flag_group].C)
+    (result, flags_out) = AddWithCarry(a, -b_shifted, -FLAGS[flag_group].C)
```

While this gives mathematically the correct arithmetic result, I find it a bit weird if we call a function `AddWithCarry` and then feed a negative number into the carry parameter.
I think, normally you would combine the borrow_in with the computation of the two's complement of `b`. The inverted borrow_in becomes the replacement for the `+1` and instead of the negation we just do an inversion of `b`.
(`c = a - b - borrow_in = a + ~b + 1 - borrow_in = a + ~b + ~borrow_in`)

Independent from this we also need to invert the carry that is generated by `AddWithCarry` for the new carry flag.

Here is an example of what i think would be the correct description. Just the concept haven't tested this.
```suggestion
    (result, flags_out) = AddWithCarry(a, b_shifted^(2**WLEN-1), ~FLAGS[flag_group].C)
    flags_out.C = ~flags_out.C
```
(The xor'ing is just the inversion but we need to let python know the width of the number, otherwise leading `1`s would be missing)

---

Good point on the carry flag.

I'd be wary of matching the RTL too closely with bit tricks to implement subtraction with an adder. I think we're better off going the other way to make things more abstract, create a `SubtractWithBorrow` function instead that does a subtraction internally (and subtracts the flag rather than adds it). The results it produces should be identical in all cases and if we've somehow messed up an edge case with the bitwise operations we're more likely to find it.

---

Sounds like a good approach to me.